Fix the remaining WPT failures for imperative slot API

rniwa opened the pull request:
#### 5473340616467e397f8dcd572c0ef84957e9ae57
<pre>
Fix the remaining WPT failures for imperative slot API
<a href="https://bugs.webkit.org/show_bug.cgi?id=243817">https://bugs.webkit.org/show_bug.cgi?id=243817</a>

Reviewed by Chris Dumez.

This patch improves the imperative slot API implementation in WebKit by adding support for slotchange event.

* LayoutTests/fast/shadow-dom/manual-assignment-multiple-shadow-roots-expected.txt: Added.
* LayoutTests/fast/shadow-dom/manual-assignment-multiple-shadow-roots.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/imperative-slot-api-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/imperative-slot-api-slotchange-expected.txt:
* LayoutTests/platform/win/TestExpectations:

* Source/WebCore/dom/Node.cpp:
(WebCore::Node::setManuallyAssignedSlot): Destruct the renderer when the assigned slot of this node changes.

* Source/WebCore/dom/ShadowRoot.cpp:
(WebCore::ShadowRoot::didRemoveManuallyAssignedNode): Added.
* Source/WebCore/dom/ShadowRoot.h:

* Source/WebCore/dom/SlotAssignment.cpp:
(WebCore::NamedSlotAssignment::didRemoveManuallyAssignedNode): Added.
(WebCore::NamedSlotAssignment::willRemoveAssignedNode):
(WebCore::ManualSlotAssignment::assignedNodesForSlot):
(WebCore::ManualSlotAssignment::addSlotElementByName): Added the logic to enqueue a slotchange event.
(WebCore::ManualSlotAssignment::removeSlotElementByName): Ditto.
(WebCore::ManualSlotAssignment::slotManualAssignmentDidChange): Now notifies slots in other shadow trees
about changes in assigned nodes to enqueue slotchange event.
(WebCore::ManualSlotAssignment::didRemoveManuallyAssignedNode): Added. Gets called when one of the assigned
nodes to re-assigned to a slot in another tree.
(WebCore::ManualSlotAssignment::willRemoveAssignedNode): Added the logic to enqueue a slotchange event.

* Source/WebCore/dom/SlotAssignment.h:
(WebCore::ShadowRoot::willRemoveAssignedNode):

* Source/WebCore/html/HTMLSlotElement.cpp:
(WebCore::HTMLSlotElement::assign): Added the logic to clear the previously assigned slot.

Canonical link: <a href="https://commits.webkit.org/253359@main">https://commits.webkit.org/253359@main</a>
</pre>

cdumez commented on Source/WebCore/dom/SlotAssignment.cpp:
```diff
@@ -453,36 +457,45 @@ void ManualSlotAssignment::renameSlotElement(HTMLSlotElement&, const AtomString&
+        shadowRoot.host()->setHasShadowRootContainingSlots(true);
+    ++m_slotElementCount;
+    ++m_slottableVersion;
+    if (auto* nowAssgined = assignedNodesForSlot(slot, shadowRoot))
```
Typo: `nowAssgined`. Also we may not need this local variable at all.

rniwa replied:
Oh yeah, I'm gonna get rid of the variable.

cdumez commented on Source/WebCore/dom/SlotAssignment.cpp:
```diff
@@ -453,36 +457,45 @@ void ManualSlotAssignment::renameSlotElement(HTMLSlotElement&, const AtomString&
-            previousSlot->removeManuallyAssignedNode(*node);
-            if (previousSlot->isInShadowTree() && previousSlot->containingShadowRoot() == slot.containingShadowRoot())
+        RefPtr protectedNode = node.get();
+        if (auto* previousSlot = protectedNode->manuallyAssignedSlot()) {
```
No RefPtr?

rniwa replied:
Good point. Fixed.

cdumez commented on Source/WebCore/dom/SlotAssignment.cpp:
```diff
@@ -530,9 +553,13 @@ void ManualSlotAssignment::hostChildElementDidChangeSlotAttribute(Element&, cons
 {
     ++m_slottableVersion;
+    if (RefPtr slot = node.assignedSlot()) {
+        if (slot->containingShadowRoot() == &shadowRoot)
```
nit: Could be on one line:
`if (RefPtr slot = node.assignedSlot(); slot->containingShadowRoot() == &shadowRoot) {`

rniwa replied:
Nice. Done.

rniwa replied:
Actually, I need to check the nullity of `slot` but that works.

cdumez replied:
Oh totally, my bad. I sometimes make this mistake :|

cdumez reviewed (approved):
LGTM

rniwa committed:
Fix the remaining WPT failures for imperative slot API
https://bugs.webkit.org/show_bug.cgi?id=243817

Reviewed by Chris Dumez.

This patch improves the imperative slot API implementation in WebKit by adding support for slotchange event.

* LayoutTests/fast/shadow-dom/manual-assignment-multiple-shadow-roots-expected.txt: Added.
* LayoutTests/fast/shadow-dom/manual-assignment-multiple-shadow-roots.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/imperative-slot-api-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/imperative-slot-api-slotchange-expected.txt:
* LayoutTests/platform/win/TestExpectations:

* Source/WebCore/dom/Node.cpp:
(WebCore::Node::setManuallyAssignedSlot): Destruct the renderer when the assigned slot of this node changes.

* Source/WebCore/dom/ShadowRoot.cpp:
(WebCore::ShadowRoot::didRemoveManuallyAssignedNode): Added.
* Source/WebCore/dom/ShadowRoot.h:

* Source/WebCore/dom/SlotAssignment.cpp:
(WebCore::NamedSlotAssignment::didRemoveManuallyAssignedNode): Added.
(WebCore::NamedSlotAssignment::willRemoveAssignedNode):
(WebCore::ManualSlotAssignment::assignedNodesForSlot):
(WebCore::ManualSlotAssignment::addSlotElementByName): Added the logic to enqueue a slotchange event.
(WebCore::ManualSlotAssignment::removeSlotElementByName): Ditto.
(WebCore::ManualSlotAssignment::slotManualAssignmentDidChange): Now notifies slots in other shadow trees
about changes in assigned nodes to enqueue slotchange event.
(WebCore::ManualSlotAssignment::didRemoveManuallyAssignedNode): Added. Gets called when one of the assigned
nodes to re-assigned to a slot in another tree.
(WebCore::ManualSlotAssignment::willRemoveAssignedNode): Added the logic to enqueue a slotchange event.

* Source/WebCore/dom/SlotAssignment.h:
(WebCore::ShadowRoot::willRemoveAssignedNode):

* Source/WebCore/html/HTMLSlotElement.cpp:
(WebCore::HTMLSlotElement::assign): Added the logic to clear the previously assigned slot.

Canonical link: https://commits.webkit.org/253359@main
